File: DevExpress-Demo/CS/LayoutControlDemo.Wpf/CarsContext.cs
using System;
using System.Windows.Media;
using System.Text;
using System.Linq;
using System.IO;
#if DXCORE3
using Microsoft.EntityFrameworkCore;
using DbModelBuilder = Microsoft.EntityFrameworkCore.ModelBuilder;
#else
using System.Data.SQLite;
using System.Data.Entity;
#endif
using System.Data.Common;
using System.ComponentModel.DataAnnotations;
using System.Collections.Generic;
using DevExpress.Internal;
using DevExpress.DemoData.Helpers;



namespace DevExpress.DemoData.Models {
    public partial class CarsContext : DbContext {
#if DXCORE3
        public CarsContext() : base() {
            SetFilePath();
            connectionString = string.Format("Data Source={0}", filePath);
        }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) {
            optionsBuilder.UseLazyLoadingProxies().UseSqlite(connectionString);
        }
        string connectionString;
#else
        public CarsContext() : base(CreateConnection(), true) { }
        public CarsContext(string connectionString) : base(connectionString) { }
        public CarsContext(DbConnection connection) : base(connection, true) { }
        
        static CarsContext() {
            Database.SetInitializer<CarsContext>(null);
        }
        static DbConnection CreateConnection() {
            SetFilePath();
            var connection = DbProviderFactories.GetFactory("System.Data.SQLite.EF6").CreateConnection();
            connection.ConnectionString = new SQLiteConnectionStringBuilder { DataSource = filePath }.ConnectionString;
            return connection;
        }
#endif
        static string filePath;
        static void SetFilePath() {
            if(filePath == null)
                filePath = DataDirectoryHelper.GetFile("cars.db", DataDirectoryHelper.DataFolderName);
            try {
                var attributes = File.GetAttributes(filePath);
                if(attributes.HasFlag(FileAttributes.ReadOnly)) {
                    File.SetAttributes(filePath, attributes & ~FileAttributes.ReadOnly);
                }
            } catch { }
        }
        public DbSet<Car> Cars { get; set; }
        public DbSet<CarScheduling> CarSchedule { get; set; }
        public DbSet<UsageType> UsageTypes { get; set; }
    }
}



namespace DevExpress.DemoData.Models {
    public class Car {
        ImageSource imageSource;
        public long Id { get; set; }
        public string Trademark { get; set; }
        public string Model { get; set; }
        public int HP { get; set; }
        public double Liter { get; set; }
        public int Cyl { get; set; }
        public int TransmissSpeedCount { get; set; }
        public string TransmissAutomatic { get; set; }
        public int MPGCity { get; set; }
        public int MPGHighway { get; set; }
        public string Category { get; set; }
        public string Description { get; set; }
        public string Hyperlink { get; set; }
        public byte[] Picture { get; set; }
        public byte[] Icon { get; set; }
        public ImageSource ImageSource {
            get {
                if(imageSource == null)
                    imageSource = ImageSourceHelper.GetImageSource(new MemoryStream(Picture));
                return imageSource;
            }
        }
        public decimal Price { get; set; }
        public DateTime DeliveryDate { get; set; }
        public bool IsInStock { get; set; }
    }

    public class CarScheduling {
        public long Id { get; set; }
        public virtual Car Car { get; set; }
        public int Status { get; set; }
        public string Subject { get; set; }
        public string Description { get; set; }
        public int Label { get; set; }
        public DateTime StartTime { get; set; }
        public DateTime EndTime { get; set; }
        public string Location { get; set; }
        public bool AllDay { get; set; }
        public int EventType { get; set; }
        public string RecurrenceInfo { get; set; }
        public string ReminderInfo { get; set; }
        public decimal Price { get; set; }
        public string ContactInfo { get; set; }
    }

    public class UsageType {
        public long Id { get; set; }
        public string Name { get; set; }
        public int Color { get; set; }
    }
}



namespace DevExpress.DemoData.Models {
    public class CarsData {
        CarsContext context = new CarsContext();
        public IEnumerable<Car> Cars {
            get { return context.Cars.ToList(); }
        }
        public IEnumerable<CarScheduling> CarsSchedule {
            get { return context.CarSchedule.ToList(); }
        }
    }
}

File: DevExpress-Demo/VB/LayoutControlDemo.Wpf/CarsContext.vb
Imports System.Windows.Media
Imports System.Linq
Imports System.IO
#If DXCORE3
using Microsoft.EntityFrameworkCore;
using DbModelBuilder = Microsoft.EntityFrameworkCore.ModelBuilder;
#Else
Imports System.Data.SQLite
Imports System.Data.Entity
#End If
Imports System.Data.Common
Imports System.ComponentModel.DataAnnotations
Imports System.Collections.Generic
Imports DevExpress.Internal
Imports DevExpress.DemoData.Helpers

Namespace DevExpress.DemoData.Models

    Public Partial Class CarsContext
        Inherits DbContext

#If DXCORE3
        public CarsContext() : base() {
            SetFilePath();
            connectionString = string.Format("Data Source={0}", filePath);
        }
        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) {
            optionsBuilder.UseLazyLoadingProxies().UseSqlite(connectionString);
        }
        string connectionString;
#Else
        Public Sub New()
            MyBase.New(CreateConnection(), True)
        End Sub

        Public Sub New(ByVal connectionString As String)
            MyBase.New(connectionString)
        End Sub

        Public Sub New(ByVal connection As DbConnection)
            MyBase.New(connection, True)
        End Sub

        Shared Sub New()
            Database.SetInitializer(Of CarsContext)(Nothing)
        End Sub

        Private Shared Function CreateConnection() As DbConnection
            Call SetFilePath()
            Dim connection = DbProviderFactories.GetFactory("System.Data.SQLite.EF6").CreateConnection()
            connection.ConnectionString = New SQLiteConnectionStringBuilder With {.DataSource = filePath}.ConnectionString
            Return connection
        End Function

#End If
        Private Shared filePath As String

        Private Shared Sub SetFilePath()
            If Equals(filePath, Nothing) Then filePath = DataDirectoryHelper.GetFile("cars.db", DataDirectoryHelper.DataFolderName)
            Try
                Dim attributes = File.GetAttributes(filePath)
                If attributes.HasFlag(FileAttributes.ReadOnly) Then
                    File.SetAttributes(filePath, attributes And Not FileAttributes.ReadOnly)
                End If
            Catch
            End Try
        End Sub

        Public Property Cars As DbSet(Of Car)

        Public Property CarSchedule As DbSet(Of CarScheduling)

        Public Property UsageTypes As DbSet(Of UsageType)
    End Class
End Namespace

Namespace DevExpress.DemoData.Models

    Public Class Car

        Private imageSourceField As ImageSource

        Public Property Id As Long

        Public Property Trademark As String

        Public Property Model As String

        Public Property HP As Integer

        Public Property Liter As Double

        Public Property Cyl As Integer

        Public Property TransmissSpeedCount As Integer

        Public Property TransmissAutomatic As String

        Public Property MPGCity As Integer

        Public Property MPGHighway As Integer

        Public Property Category As String

        Public Property Description As String

        Public Property Hyperlink As String

        Public Property Picture As Byte()

        Public Property Icon As Byte()

        Public ReadOnly Property ImageSource As ImageSource
            Get
                If imageSourceField Is Nothing Then imageSourceField = ImageSourceHelper.GetImageSource(New MemoryStream(Picture))
                Return imageSourceField
            End Get
        End Property

        Public Property Price As Decimal

        Public Property DeliveryDate As Date

        Public Property IsInStock As Boolean
    End Class

    Public Class CarScheduling

        Public Property Id As Long

        Public Overridable Property Car As Car

        Public Property Status As Integer

        Public Property Subject As String

        Public Property Description As String

        Public Property Label As Integer

        Public Property StartTime As Date

        Public Property EndTime As Date

        Public Property Location As String

        Public Property AllDay As Boolean

        Public Property EventType As Integer

        Public Property RecurrenceInfo As String

        Public Property ReminderInfo As String

        Public Property Price As Decimal

        Public Property ContactInfo As String
    End Class

    Public Class UsageType

        Public Property Id As Long

        Public Property Name As String

        Public Property Color As Integer
    End Class
End Namespace

Namespace DevExpress.DemoData.Models

    Public Class CarsData

        Private context As CarsContext = New CarsContext()

        Public ReadOnly Property Cars As IEnumerable(Of Car)
            Get
                Return context.Cars.ToList()
            End Get
        End Property

        Public ReadOnly Property CarsSchedule As IEnumerable(Of CarScheduling)
            Get
                Return context.CarSchedule.ToList()
            End Get
        End Property
    End Class
End Namespace
